File: src/types/pitstop.ts
/**
 * Type definitions for Pitstop API responses.
 */

/** Backend job status values */
export type PitstopJobStatus = "QUEUED" | "PROCESSING" | "COMPLETE" | "FAILED";

/** Backend processing stages */
export type PitstopStage = 
  | "UPLOAD" 
  | "DETECTING" 
  | "TRACKING" 
  | "RENDERING" 
  | "COMPLETE" 
  | "FAILED";

/** Output file metadata from backend */
export interface PitstopOutput {
  available: boolean;
  filename?: string | null;
  size_bytes?: number | null;
}

/** Full job response from GET /api/pitstop/jobs/{job_id} */
export interface PitstopJob {
  job_id: string;
  status: PitstopJobStatus;
  stage?: string;
  progress?: number;
  series?: string | null;
  race?: string | null;
  notes?: string | null;
  input_filename?: string;
  input_size_bytes?: number;
  logs?: string[];
  output?: PitstopOutput;
  created_at?: string;
  updated_at?: string;
}

/** Response from POST /api/pitstop/jobs */
export interface CreateJobResponse extends PitstopJob {}

/** Metadata for job creation */
export interface JobMetadata {
  series?: string;
  race?: string;
  notes?: string;
}

/** Lightweight job item for list endpoint (GET /api/pitstop/jobs) */
export interface PitstopJobListItem {
  job_id: string;
  status: PitstopJobStatus;
  stage: string;
  progress: number;
  series?: string | null;
  race?: string | null;
  input_filename: string;
  input_size_bytes: number;
  has_output: boolean;
  created_at: string;
  updated_at: string;
}

/** Response from GET /api/pitstop/jobs with pagination metadata */
export interface PitstopJobListResponse {
  items: PitstopJobListItem[];
  total: number;
  limit: number;
  offset: number;
}

/** Timing metrics for a pitstop job from GET /api/pitstop/jobs/{job_id}/metrics */
export interface PitstopRunMetrics {
  job_id: string;
  fuel_time_s: number | null;
  front_left_tyre_time_s: number | null;
  front_right_tyre_time_s: number | null;
  back_left_tyre_time_s: number | null;
  back_right_tyre_time_s: number | null;
  driver_out_time_s: number | null;
  driver_in_time_s: number | null;
}

/** UI-level job status (combines backend status with upload phase) */
export type UIJobStatus = 
  | "idle" 
  | "uploading" 
  | "queued" 
  | "processing" 
  | "rendering" 
  | "complete" 
  | "failed";

/**
 * Map backend status + stage to UI status.
 */
export function mapBackendToUIStatus(
  status: PitstopJobStatus | undefined,
  stage: string | undefined
): UIJobStatus {
  if (!status) return "idle";
  
  switch (status) {
    case "QUEUED":
      return "queued";
    case "PROCESSING":
      // Map stage to more granular UI status
      if (stage === "RENDERING") return "rendering";
      return "processing";
    case "COMPLETE":
      return "complete";
    case "FAILED":
      return "failed";
    default:
      return "idle";
  }
}

/**
 * Get the active step index for the stepper based on backend status/stage.
 * Steps: Upload(0), Queued(1), Processing(2), Rendering(3), Complete(4)
 */
export function getStepFromBackend(
  status: PitstopJobStatus | undefined,
  stage: string | undefined
): number {
  if (!status) return -1;
  
  switch (status) {
    case "QUEUED":
      if (stage === "UPLOAD") return 0;
      return 1;
    case "PROCESSING":
      if (stage === "DETECTING" || stage === "TRACKING") return 2;
      if (stage === "RENDERING") return 3;
      return 2;
    case "COMPLETE":
      return 5; // Past the last step to show all complete
    case "FAILED":
      return -1;
    default:
      return -1;
  }
}

File: backend/app/schemas/pitstop.py
from __future__ import annotations

from datetime import datetime
from typing import List, Optional
from uuid import UUID

from pydantic import BaseModel, Field

from app.db.models import JobStatus


class OutputInfo(BaseModel):
    available: bool
    filename: Optional[str] = None
    size_bytes: Optional[int] = None


class PitstopJobCreate(BaseModel):
    """Request schema - used internally, file comes from form."""

    series: Optional[str] = None
    race: Optional[str] = None
    notes: Optional[str] = None


class PitstopJobResponse(BaseModel):
    """Response schema for job creation and status."""

    job_id: UUID
    status: JobStatus
    stage: str
    progress: float = Field(ge=0.0, le=1.0)
    mode: str = "classic"
    series: Optional[str]
    race: Optional[str]
    notes: Optional[str]
    input_filename: str
    input_size_bytes: int
    logs: List[str] = []
    output: OutputInfo
    error_message: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}

    @classmethod
    def from_job(cls, job) -> "PitstopJobResponse":
        """Create response from PitstopJob model."""
        output = OutputInfo(
            available=job.status == JobStatus.COMPLETE and job.output_path is not None,
            filename=job.output_filename,
            size_bytes=job.output_size_bytes,
        )
        return cls(
            job_id=job.id,
            status=job.status,
            stage=job.stage,
            progress=job.progress,
            mode=getattr(job, 'mode', 'classic'),
            series=job.series,
            race=job.race,
            notes=job.notes,
            input_filename=job.input_filename,
            input_size_bytes=job.input_size_bytes,
            logs=job.get_logs_tail(200),
            output=output,
            error_message=getattr(job, 'error_message', None),
            created_at=job.created_at,
            updated_at=job.updated_at,
        )


class PitstopJobListItem(BaseModel):
    """Lightweight schema for job listing."""

    job_id: UUID
    status: JobStatus
    stage: str
    progress: float
    mode: str = "classic"
    series: Optional[str]
    race: Optional[str]
    input_filename: str
    input_size_bytes: int
    has_output: bool
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}

    @classmethod
    def from_job(cls, job) -> "PitstopJobListItem":
        return cls(
            job_id=job.id,
            status=job.status,
            stage=job.stage,
            progress=job.progress,
            mode=getattr(job, 'mode', 'classic'),
            series=job.series,
            race=job.race,
            input_filename=job.input_filename,
            input_size_bytes=job.input_size_bytes,
            has_output=job.output_path is not None,
            created_at=job.created_at,
            updated_at=job.updated_at,
        )


class PitstopJobListResponse(BaseModel):
    """Response for job listing endpoint with pagination metadata."""

    items: List[PitstopJobListItem]
    total: int
    limit: int
    offset: int


class PitstopMetricsUpdate(BaseModel):
    """Request schema for updating job metrics (POST).
    
    All fields are optional - only provided fields will be updated.
    Use this endpoint to manually set metrics for testing UI.
    """
    
    fuel_time_s: Optional[float] = Field(None, ge=0, description="Fuel time in seconds")
    front_left_tyre_time_s: Optional[float] = Field(None, ge=0, description="Front left tyre change time")
    front_right_tyre_time_s: Optional[float] = Field(None, ge=0, description="Front right tyre change time")
    back_left_tyre_time_s: Optional[float] = Field(None, ge=0, description="Back left tyre change time")
    back_right_tyre_time_s: Optional[float] = Field(None, ge=0, description="Back right tyre change time")
    driver_out_time_s: Optional[float] = Field(None, ge=0, description="Driver exit time")
    driver_in_time_s: Optional[float] = Field(None, ge=0, description="Driver entry time")
    
    def to_payload(self) -> dict:
        """Convert to dict with only non-None values for upsert."""
        return {k: v for k, v in self.model_dump().items() if v is not None}


class PitstopRunMetricsOut(BaseModel):
    """Response schema for run metrics endpoint.
    
    Note: Uses job_id to match the updated schema (previously run_id).
    """

    job_id: UUID
    fuel_time_s: Optional[float] = None
    front_left_tyre_time_s: Optional[float] = None
    front_right_tyre_time_s: Optional[float] = None
    back_left_tyre_time_s: Optional[float] = None
    back_right_tyre_time_s: Optional[float] = None
    driver_out_time_s: Optional[float] = None
    driver_in_time_s: Optional[float] = None

    model_config = {"from_attributes": True}

    @classmethod
    def from_summary(cls, summary, job_id: UUID) -> "PitstopRunMetricsOut":
        """Create response from PitstopBreakdownSummary model or return empty metrics."""
        if summary is None:
            return cls(job_id=job_id)
        return cls(
            job_id=summary.job_id,
            fuel_time_s=summary.fuel_time_s,
            front_left_tyre_time_s=summary.front_left_tyre_time_s,
            front_right_tyre_time_s=summary.front_right_tyre_time_s,
            back_left_tyre_time_s=summary.back_left_tyre_time_s,
            back_right_tyre_time_s=summary.back_right_tyre_time_s,
            driver_out_time_s=summary.driver_out_time_s,
            driver_in_time_s=summary.driver_in_time_s,
        )
